AnnService/inc/Core/Common/Labelset.h

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#ifndef _SPTAG_COMMON_LABELSET_H_
+#define _SPTAG_COMMON_LABELSET_H_
+
+#include <atomic>
+#include <shared_mutex>
+#include "Dataset.h"
+
+namespace SPTAG
+{
+    namespace COMMON
+    {
+        class Labelset
+        {
+        private:
+            std::atomic<SizeType> m_deleted;
+            Dataset<std::int8_t> m_data;
+            std::unique_ptr<std::shared_timed_mutex> m_lock;
+
+        public:
+            Labelset() 
+            {
+                m_deleted = 0;
+                m_data.SetName("DeleteID");
+                m_lock.reset(new std::shared_timed_mutex);
+            }
+
+            void Initialize(SizeType capacity)
+            {
+                m_data.Initialize(capacity, 1);
+            }
+
+            inline size_t Size() const { return m_deleted.load(); }
+
+            inline bool Contains(const SizeType& key) const
+            {
+                return *m_data[key] == 1;
+            }
+
+            inline void Insert(const SizeType& key)
+            {
+                if (*m_data[key] != 1)
+                {
+                    std::shared_lock<std::shared_timed_mutex> lock(*m_lock);
+                    m_deleted++;
```

Comment:
should check if m_data[key] != 1 after locked

Reply:
This lock prevents any deletes when refining index. It does not prevent two same deletes at the same time.